extensions/kubernetes/controllerservice/KubernetesControllerService.cpp

```diff
@@ -0,0 +1,205 @@
+
+KubernetesControllerService::KubernetesControllerService(const std::string& name, const utils::Identifier& uuid)
+  : AttributeProviderService(name, uuid),
+    logger_{core::logging::LoggerFactory<KubernetesControllerService>::getLogger()},
+    api_client_{std::make_unique<APIClient>(*logger_)} {
+}
+
+KubernetesControllerService::KubernetesControllerService(const std::string& name, const std::shared_ptr<Configure>& configuration)
+  : KubernetesControllerService{name} {
+    setConfiguration(configuration);
+    initialize();
+}
+
+void KubernetesControllerService::initialize() {
+  if (initialized_) { return; }
+
+  std::lock_guard<std::mutex> lock(initialization_mutex_);
+  ControllerService::initialize();
+  setSupportedProperties({NamespaceFilter, PodNameFilter, ContainerNameFilter});
+  initialized_ = true;
+}
+
+void KubernetesControllerService::onEnable() {
+  std::string namespace_filter;
+  if (getProperty(NamespaceFilter.getName(), namespace_filter) && !namespace_filter.empty()) {
+    namespace_filter_ = std::regex{namespace_filter};
+  }
+
+  std::string pod_name_filter;
+  if (getProperty(PodNameFilter.getName(), pod_name_filter) && !pod_name_filter.empty()) {
+    pod_name_filter_ = std::regex{pod_name_filter};
+  }
+
+  std::string container_name_filter;
+  if (getProperty(ContainerNameFilter.getName(), container_name_filter) && !container_name_filter.empty()) {
+    container_name_filter_ = std::regex{container_name_filter};
+  }
+}
+
+namespace {
+
+struct v1_pod_list_t_deleter {
+  void operator()(v1_pod_list_t* ptr) const noexcept { v1_pod_list_free(ptr); }
+};
+using v1_pod_list_unique_ptr = std::unique_ptr<v1_pod_list_t, v1_pod_list_t_deleter>;
+
+v1_pod_list_unique_ptr getPods(apiClient_t* api_client, core::logging::Logger& logger) {
+  logger.log_info("Calling Kubernetes API listPodForAllNamespaces...");
+  v1_pod_list_unique_ptr pod_list{CoreV1API_listPodForAllNamespaces(api_client,
+                                                                    0,  // allowWatchBookmarks
+                                                                    nullptr,  // continue
+                                                                    nullptr,  // fieldSelector
+                                                                    nullptr,  // labelSelector
+                                                                    0,  // limit
+                                                                    nullptr,  // pretty
+                                                                    nullptr,  // resourceVersion
+                                                                    nullptr,  // resourceVersionMatch
+                                                                    0,  // timeoutSeconds
+                                                                    0)};  // watch
+  logger.log_info("The return code of the Kubernetes API listPodForAllNamespaces call: %ld", api_client->response_code);
```

Comment:
The way `CoreV1API_listPodForAllNamespaces` signals an error is by returning a null pointer.  I could change the return value of `getPods()` to include the HTTP response code in case of an error, but that would complicate the code, and the caller can't do much with the error code than logging it, either.

Since this is a local function which is not visible outside the file, I don't think it's worth doing that.

Reply:
I see, let's keep it this way then.